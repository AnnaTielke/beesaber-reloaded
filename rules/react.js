module.exports = {
  plugins: ['react'],

  extends: ['plugin:react/recommended'],

  settings: {
    react: {
      version: 'detect',
    },
  },

  rules: {
    'react/button-has-type': 'error',
    'react/jsx-boolean-value': 'error',
    'react/jsx-fragments': 'error',
    'react/jsx-handler-names': 'error',
    'react/jsx-no-bind': ['error', { allowArrowFunctions: true }],
    'react/jsx-pascal-case': 'error',
    'react/no-access-state-in-setstate': 'error',
    'react/no-array-index-key': 'error',
    'react/no-danger': 'error',
    'react/no-did-mount-set-state': 'error',
    'react/no-did-update-set-state': 'error',
    'react/no-multi-comp': 'error',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unsafe': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/no-will-update-set-state': 'error',
    'react/prefer-es6-class': 'error',
    'react/prefer-stateless-function': 'error',
    'react/style-prop-object': 'error',
    'react/void-dom-elements-no-children': 'error',
  },
};
