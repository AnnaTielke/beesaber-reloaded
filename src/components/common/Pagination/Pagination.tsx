import { useEffect, useState } from 'react';
// import '../styles/Pagination.scss';

type Page = number | string;

type PaginationData = {
  currentPage: number;
  pageLimit: number;
  totalPages: number;
  totalScores: number;
};

type PaginationProps = Pick<PaginationData, 'pageLimit' | 'totalScores'> & {
  onPageChanged: (paginationData: PaginationData) => void;
  pageNeighbours?: number;
};

type FetchPageNumbersProps = Required<
  Pick<PaginationData, 'currentPage' | 'totalPages'>
> &
  Required<Pick<PaginationProps, 'pageNeighbours'>>;

const LEFT_PAGE = 'LEFT';
const RIGHT_PAGE = 'RIGHT';

const range = (from: number, to: number, step = 1): number[] => {
  let i = from;
  const rangeArr: number[] = [];

  while (i <= to) {
    rangeArr.push(i);
    i += step;
  }

  return rangeArr;
};

const fetchPageNumbers = ({
  currentPage,
  pageNeighbours,
  totalPages,
}: FetchPageNumbersProps) => {
  const totalNumbers = pageNeighbours * 2 + 3;
  const totalBlocks = totalNumbers + 2;

  if (totalPages > totalBlocks) {
    let pages = [];

    const leftBound = currentPage - pageNeighbours;
    const rightBound = currentPage + pageNeighbours;
    const beforeLastPage = totalPages - 1;

    const startPage = leftBound > 2 ? leftBound : 2;
    const endPage = rightBound < beforeLastPage ? rightBound : beforeLastPage;

    pages = range(startPage, endPage);

    const pagesCount = pages.length;
    const singleSpillOffset = totalNumbers - pagesCount - 1;

    const leftSpill = startPage > 2;
    const rightSpill = endPage < beforeLastPage;

    const leftSpillPage = LEFT_PAGE;
    const rightSpillPage = RIGHT_PAGE;

    if (leftSpill && !rightSpill) {
      const extraPages = range(startPage - singleSpillOffset, startPage - 1);

      pages = [leftSpillPage, ...extraPages, ...pages];
    } else if (!leftSpill && rightSpill) {
      const extraPages = range(endPage + 1, endPage + singleSpillOffset);

      pages = [...pages, ...extraPages, rightSpillPage];
    } else if (leftSpill && rightSpill) {
      pages = [leftSpillPage, ...pages, rightSpillPage];
    }

    return [1, ...pages, totalPages];
  }

  return range(1, totalPages);
};

const Pagination = ({
  onPageChanged,
  pageLimit = 30,
  pageNeighbours = 0,
  totalScores,
}: PaginationProps): JSX.Element | null => {
  const [currentPage, setCurrentPage] = useState<number>(1);
  const [totalPages, setTotalPages] = useState<number>(
    Math.ceil(totalScores / pageLimit)
  );

  // pageNeighbours can be: 0, 1 or 2
  pageNeighbours = Math.max(0, Math.min(pageNeighbours, 2));

  const pages = fetchPageNumbers({ currentPage, pageNeighbours, totalPages });

  const gotoPage = (page: Page): void => {
    const newCurrentPage = Math.max(0, Math.min(Number(page), totalPages));
    const paginationData = {
      currentPage,
      totalPages,
      pageLimit,
      totalScores,
    };

    onPageChanged(paginationData);

    setCurrentPage(newCurrentPage);
  };

  const handleClick = (
    page: Page | string,
    evt: React.MouseEvent<HTMLElement>
  ) => {
    evt.preventDefault();
    gotoPage(page);
  };

  const handleMoveLeft = (evt: React.MouseEvent<HTMLElement>) => {
    evt.preventDefault();
    gotoPage(currentPage - pageNeighbours * 2 - 1);
  };

  const handleMoveRight = (evt: React.MouseEvent<HTMLElement>) => {
    evt.preventDefault();
    gotoPage(currentPage + pageNeighbours * 2 + 1);
  };

  useEffect(() => {
    gotoPage(1);
  }, []);

  useEffect(() => {
    setTotalPages(Math.ceil(totalScores / pageLimit));
  }, [totalScores]);

  if (totalPages < 2) return null;

  return (
    <div id="pagination">
      <>
        <nav aria-label="Scores Pagination">
          <ul className="pagination">
            {pages.map((page, index) => {
              if (page === LEFT_PAGE)
                return (
                  <li key={index} className="page-item">
                    <a
                      className="page-link"
                      href="/#"
                      aria-label="Previous"
                      onClick={handleMoveLeft}
                    >
                      <span aria-hidden="true">&laquo;</span>
                      <span className="sr-only">Previous</span>
                    </a>
                  </li>
                );

              if (page === RIGHT_PAGE)
                return (
                  <li key={index} className="page-item">
                    <a
                      className="page-link"
                      href="/#"
                      aria-label="Next"
                      onClick={handleMoveRight}
                    >
                      <span aria-hidden="true">&raquo;</span>
                      <span className="sr-only">Next</span>
                    </a>
                  </li>
                );

              return (
                <li
                  key={index}
                  className={`page-item${
                    currentPage === page ? ' active' : ''
                  }`}
                >
                  <a
                    className="page-link"
                    href="/#"
                    onClick={(e) => handleClick(page, e)}
                  >
                    {page}
                  </a>
                </li>
              );
            })}
          </ul>
        </nav>
      </>
    </div>
  );
};

export default Pagination;
