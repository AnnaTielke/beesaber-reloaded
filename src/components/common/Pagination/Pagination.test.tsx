import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Pagination from './Pagination';

describe('Pagination', () => {
  const mockOnPageChange = jest.fn();

  afterEach(() => jest.resetAllMocks());

  it('should match the snapshot on initial view', () => {
    const { container } = render(
      <Pagination
        onPageChanged={mockOnPageChange}
        pageLimit={5}
        totalScores={30}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match the snapshot after selecting a page', () => {
    const { container } = render(
      <Pagination
        onPageChanged={mockOnPageChange}
        pageLimit={5}
        totalScores={30}
      />
    );

    const pageLink = screen.getByRole('link', { name: '3' });

    userEvent.click(pageLink);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render, when totalpages < 2', () => {
    const { container } = render(
      <Pagination
        onPageChanged={mockOnPageChange}
        pageLimit={30}
        totalScores={10}
      />
    );

    expect(container).toBeEmptyDOMElement();
  });

  it('should render, when totalpages >= 2', () => {
    const { container } = render(
      <Pagination
        onPageChanged={mockOnPageChange}
        pageLimit={5}
        totalScores={10}
      />
    );

    expect(container).not.toBeEmptyDOMElement();
  });

  it('should call onPageChanged when selecting a page', () => {
    render(
      <Pagination
        onPageChanged={mockOnPageChange}
        pageLimit={5}
        totalScores={10}
      />
    );

    const pageLink = screen.getByRole('link', { name: '1' });

    userEvent.click(pageLink);

    expect(mockOnPageChange).toHaveBeenCalledTimes(1);
    expect(mockOnPageChange).toHaveBeenCalledWith({
      currentPage: 1,
      totalPages: 2,
      pageLimit: 5,
      totalScores: 10,
    });
  });

  it.each`
    direction  | text          | previousPage | nextPage
    ${'Left'}  | ${'Previous'} | ${3}         | ${2}
    ${'Right'} | ${'Next'}     | ${3}         | ${4}
  `(
    'should call handleMove$direction when clicking $direction arrow',
    ({ text, previousPage, nextPage }) => {
      render(
        <Pagination
          onPageChanged={mockOnPageChange}
          pageLimit={5}
          totalScores={50}
        />
      );

      const pageLink = screen.getByRole('link', { name: '3' });

      userEvent.click(pageLink);

      const nextLink = screen.getByRole('link', { name: text });

      userEvent.click(nextLink);

      expect(mockOnPageChange).toHaveBeenCalledTimes(2);
      expect(JSON.stringify(mockOnPageChange.mock.calls)).toStrictEqual(
        JSON.stringify([
          [
            {
              currentPage: previousPage,
              totalPages: 10,
              pageLimit: 5,
              totalScores: 50,
            },
          ],
          [
            {
              currentPage: nextPage,
              totalPages: 10,
              pageLimit: 5,
              totalScores: 50,
            },
          ],
        ])
      );
    }
  );
});
