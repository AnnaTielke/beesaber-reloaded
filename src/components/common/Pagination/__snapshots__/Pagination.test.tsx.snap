// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match the snapshot after selecting a page 1`] = `
.c3 {
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c4 {
  color: #00111f;
  padding: 0.3rem;
}

.c5 {
  background-color: #ffeb00;
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c1 {
  background-color: rgba(255,255,255,0.7);
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c2 {
  color: #00111f;
  padding: 0.3rem;
}

.c0 {
  box-shadow: 0 0 5px rgba(0,0,0,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  list-style-type: none;
  margin-bottom: 0;
  margin-top: 0;
}

<nav
  class=""
>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        class="c2"
        href="/#"
      >
        1
      </a>
    </li>
    <li
      class="c3"
    >
      <a
        aria-label="Previous"
        class="c4"
        href="/#"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span>
          Previous
        </span>
      </a>
    </li>
    <li
      class="c5"
    >
      <a
        class="c2"
        href="/#"
      >
        3
      </a>
    </li>
    <li
      class="c3"
    >
      <a
        aria-label="Next"
        class="c4"
        href="/#"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span>
          Next
        </span>
      </a>
    </li>
    <li
      class="c1"
    >
      <a
        class="c2"
        href="/#"
      >
        6
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should match the snapshot on initial view 1`] = `
.c4 {
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c5 {
  color: #00111f;
  padding: 0.3rem;
}

.c1 {
  background-color: #ffeb00;
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c3 {
  background-color: rgba(255,255,255,0.7);
  font-size: 1rem;
  font-weight: 900;
  padding: 0.3rem;
  text-align: center;
}

.c2 {
  color: #00111f;
  padding: 0.3rem;
}

.c0 {
  box-shadow: 0 0 5px rgba(0,0,0,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  list-style-type: none;
  margin-bottom: 0;
  margin-top: 0;
}

<nav
  class=""
>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <a
        class="c2"
        href="/#"
      >
        1
      </a>
    </li>
    <li
      class="c3"
    >
      <a
        class="c2"
        href="/#"
      >
        2
      </a>
    </li>
    <li
      class="c3"
    >
      <a
        class="c2"
        href="/#"
      >
        3
      </a>
    </li>
    <li
      class="c4"
    >
      <a
        aria-label="Next"
        class="c5"
        href="/#"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span>
          Next
        </span>
      </a>
    </li>
    <li
      class="c3"
    >
      <a
        class="c2"
        href="/#"
      >
        6
      </a>
    </li>
  </ul>
</nav>
`;
